### PR TITLE
Create new step for `run bundle` to install an operator bundle

### PR DESCRIPTION
**Pull request by rashmigottipati**
Signed-off-by: Rashmi Gottipati <chowdary.grashmi@gmail.com>

### CODE REVIEW

File: ci-operator/step-registry/optional-operators/ci/subscribe/optional-operators-ci-subscribe-run-bundle.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+ref:
+  as: optional-operators-ci-run-bundle
+  from: cli
+  commands: optional-operators-ci-run-bundle-commands.sh
```
Comment (jmrodri) [inline ci-operator/step-registry/optional-operators/ci/subscribe/optional-operators-ci-subscribe-run-bundle.yaml:4]: Does this still need to be a shell script? Can't this be `operator-sdk run bundle ... `?

Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/subscribe/optional-operators-ci-subscribe-run-bundle.yaml]: It has to be a shell script I think as it will be injected into the image.

File: ci-operator/step-registry/optional-operators/subscribe/optional-operators-subscribe-commands.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@ REHEARSAL_INDEX="brew.registry.redhat.io/rh-osbs-stage/iib:23576"
 REHEARSAL_PACKAGE="e2e-test-operator"
 REHEARSAL_CHANNEL="4.3"
 REHEARSAL_INSTALL_NAMESPACE="!create"
-REHEARSAL_TARGET_NAMESPACES="!install"
+REHEARSAL_TARGET_NAMESPACES="!all"
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/subscribe/optional-operators-subscribe-commands.sh:13]: I don't think you need to touch this: this triggers some special jobs to be rehearsed which is not desirable.

Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/subscribe/optional-operators-subscribe-commands.sh]: Sounds good, fixed it.

File: ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml [hunk 3]
```diff
@@ -37,11 +38,14 @@ tests:
...
+      OO_BUNDLE_IMG: nmo-bundle
     env:
-      OO_CHANNEL: "4.8"
       OO_INSTALL_NAMESPACE: default
-      OO_PACKAGE: node-maintenance-operator
-      OO_TARGET_NAMESPACES: '!install'
+      OO_TARGET_NAMESPACES: OwnNamespace
+    pre:
+    - chain: ipi-aws-pre
+    - ref: optional-operators-ci-run-bundle
```
Comment (petr-muller) [inline ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml:48]: Please remove the modification of the existing job: we only use that one for testing, but we definitely do not want  to actually mess with someone's CI jobs.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+echo "Deploying an operator in the bundle format using operator-sdk run bundle command"
+
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:8]: `operator-sdk` only supports certain install modes, so it would be useful if the step validated the parameter contains one of the supported falues and failed fast with a reasonable message if it does not.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+cd /tmp
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:9]: Why is this needed?

**Review by petr-muller — COMMENTED** [review]
We need real `OWNERS` for this step

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+operator-sdk run bundle "${OO_BUNDLE_IMG}" \
+                        --index-image "${OO_INDEX_IMG}" \
+                        --namespace "${OO_INSTALL_NAMESPACE}" \
+                        --install-mode "${OO_TARGET_NAMESPACES}"
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:13]: Can this fail? If it does, do we need to collect some artifacts for debugging?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      cpu: 300m
+      memory: 300Mi
+  dependencies:
+  - name: bundle-image
+    env: OO_BUNDLE_IMG
+  env:
+  - name: OO_INSTALL_NAMESPACE
+    default: "!create"
+    documentation: |
+      This namespace must already exist in the cluster.
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:19]: This is not useful in a context of a CI job - that's the reason why the other step supported the special `!create` value. IMO, this step needs to support it too.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      This is an optional field which will default to the kubeconfig context if not provided.
+  - name: OO_TARGET_NAMESPACES
+    default: "!all"
+    documentation: |
+      The default value `!all` means the Operator will watch all namespaces (cluster-scoped Operators).
+      If the value is `!own`, then the Operator will watch its own namespace (from namespace or the kubeconfig default).
+      If the value is `!single`, the the Operator will watch a namespace, not necessarily its own.
+      This is an optional parameter, but if the CSV does not support AllNamespaces,
+      then this parameter becomes required to instruct run bundle with the appropriate target namespace.
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:28]: Looks like the documentation is misleading here: what interprets the `!all/!own/!single` values? You seem to pass the value of this param straight to `run bundle`.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  - name: OO_INDEX_IMG
+    default: "quay.io/operator-framework/upstream-opm-builder:latest"
+    documentation: |
+      This specifies an index image in which to inject the given bundle.
+      This is an optional field which will default to quay.io/operator-framework/upstream-opm-builder:latest
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:33]: No need to specify the parameter is optional (thats implied by it having a default value) nor to duplicate it default value (there's a structured field for that).

File: ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml (hunk 3 above)
Comment (rashmigottipati) [reply ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml:48]: Would it be ok to leave it here for now for testing purposes and remove the commit that changes the CI job before merging the PR?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh (hunk 4 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:8]: Fixed it.

File: ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml (hunk 3 above)
Comment (rashmigottipati) [reply ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml:48]: Yes, absolutely. Would it be ok to keep this for testing purposes and remove the modifications before merging the PR?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh (hunk 4 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:8]: Fixed it.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml (hunk 8 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:27]: Fixed the doc. So the supported values are `AllNamespaces`, `OwnNamespace` (being the default), and `SingleNamespace`.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh (hunk 6 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:19]: When `run bundle` is pulling the bundle image and extracting its contents, it would write those to ROOT, which would cause a permission denied error. So just changing the working directory to `tmp` instead.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml (hunk 7 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:19]: Are you suggesting that the step should support a namespace doesn't already exist in the cluster?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh (hunk 5 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:15]: When run bundle is pulling the bundle image and extracting its contents, it would write those to ROOT, which would cause a permission denied error. So just changing the working directory to tmp instead.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml (hunk 7 above)
Comment (petr-muller) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:19]: Yes. Exactly like the other step does. It's a common testing scenario: the operator needs to be installed in *some* namespace, but the CI job author does not care which one, so the other step implemented the `!create` special value for this.

File: ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml (hunk 3 above)
Comment (petr-muller) [reply ci-operator/config/kubevirt/node-maintenance-operator/kubevirt-node-maintenance-operator-master.yaml:48]: Sure, we can keep the change until we're happy with the PR and only remove it just before merge.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh (hunk 6 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:33]: Added collection of artifacts after `run bundle` is executed.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml (hunk 7 above)
Comment (rashmigottipati) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml]: The step now handles creation of namespace, similar to the `ci-subscribe` operator.

**Review by petr-muller — COMMENTED** [review]
The step needs non-empty OWNERS

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  resources:
+    requests:
+      cpu: 300m
+      memory: 300Mi
+  dependencies:
+  - name: bundle-image
+    env: OO_BUNDLE_IMG
+  env:
+  - name: OO_INSTALL_NAMESPACE
+    default: "create"
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:17]: Let's be consistent with the other step and prefix the special value with an exclamation mark. For consistency's sake, and also to allow actually calling a namespace `create` ;)

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+# create the operator bundle using `run bundle` command.
+operator-sdk run bundle "${OO_BUNDLE_IMG}" \
+                        --index-image "${OO_INDEX_IMG}" \
+                        --namespace "${OO_INSTALL_NAMESPACE}" \
+                        --install-mode "${OO_TARGET_NAMESPACES}"
+
+# create artifacts directory to collect all the manifests in yaml format.
+ARTIFACT_DIR=$TMPDIR/artifacts
+mkdir $ARTIFACT_DIR && cd $ARTIFACT_DIR
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:37]: All these will only run when `operator-sdk run buncle` succeeds (b/c the `set -o errexit` at L4), is that okay? I guess this collection will be mostly useful when the `run bundle` fails.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 12]
```diff
@@ -0,0 +1,35 @@
+ref:
+  as: optional-operators-ci-run-bundle
+  from_image:
+    namespace: ci
+    name: cvp-operator-scorecard
+    tag: v1
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:6]: I don't think we want this step to depend on CVP images. What image would typically contain operator-sdk binary?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    documentation: |
+      The namespace into which the operator and catalog will be installed.
+      Special, default value `create` implies that a new namespace will be created by the step.
+  - name: OO_TARGET_NAMESPACES
+    default: "OwnNamespace"
+    documentation: |
+      The default value `OwnNamespace` means that the Operator will watch its own namespace (from namespace or the kubeconfig default).
+      If the value is `AllNamespaces`, then the Operator will watch all namespaces (cluster-scoped Operators).
+      If the value is `SingleNamespace`, then the Operator will watch a namespace, not necessarily its own.
+      This is an optional parameter, but if the CSV does not support AllNamespaces,
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:27]: Looks like this sentence is now outdated - it says `AllNamespaces` but the default is `OwnNamespace`. How about not discussing this in the context of being mandatory/optional parameter, but rather pointing out that it's the users responsibility to set this to a value that their operator actually supports? Related to that: maybe this should not even have a default value and so force the user to intentionally specify something? That would sound better to me if the default value is not guaranteed to always work.

What can the user expect when tthey pass an unsupported value? Does `operator-sdk run bundle` explicitly point that out in a fialure, or will the user observe something like a timeout?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.metadata.json [hunk 14]
```diff
@@ -0,0 +1,8 @@
+{
+	"path": "optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml",
+	"owners": {
+		"approvers": [
+			"dptp"
```
Comment (petr-muller) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.metadata.json:5]: Nice try ;)

This goes to `OWNERS` file (ci-operator/step-registry/optional-operators/ci/run-bundle/OWNERS) (and the metadata is generated from it).

Second, DPTP will not own this step. It was our understanding that the SDK team would maintain this step for their users (operator authors): this is a principle used across the whole OpenShift CI. For example, OCP installer team maintans the steps that iinstall OCP and are used across the org.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh [hunk 15]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+echo "Deploying an operator in the bundle format using operator-sdk run bundle command"
+
+# `run bundle` supports three install modes: AllNamespaces, OwnNamespace and SingleNamespace.
+if [[ "${OO_TARGET_NAMESPACES}" != "AllNamespaces" && "${OO_TARGET_NAMESPACES}" != "OwnNamespace" && "${OO_TARGET_NAMESPACES}" != "SingleNamespace" ]]; then
```
Comment (estroz) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:10]: FYI `run bundle` will handle this check. It does support `MultiNamespace` so that should still be checked here.

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    echo "install mode should be one of the supported install modes {AllNamespaces, OwnNamespace, SingleNamespace}"
+    exit 1
+fi
+
+TMPDIR=/tmp
+cd $TMPDIR
+
+# handle creation of install namespace if it does not exist
+if [[ "$OO_INSTALL_NAMESPACE" == "create" ]]; then
```
Comment (estroz) [inline ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:19]: Should this be
```suggestion
if [[ "$OO_INSTALL_NAMESPACE" == "!create" ]]; then
```
?

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh (hunk 11 above)
Comment (estroz) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-commands.sh:37]: Yes, everything from here down should be wrapped in a function that gets called in a trap at the top of the script
```sh
set -o nounset
set -o errexit
set -o errtrace # Needed  for function to inherit the trap.
set -o pipefail

function log_resources() {

	POD_ART="$ARTIFACT_DIR/registrypod.yaml"
    echo "Logging Registry pod as $POD_ART"

	...
}

trap 'log_resources' ERR

echo "Deploying an operator in the bundle format using operator-sdk run bundle command"

...
```

File: ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  - name: OO_TARGET_NAMESPACES
+    default: "OwnNamespace"
+    documentation: |
+      The default value `OwnNamespace` means that the Operator will watch its own namespace (from namespace or the kubeconfig default).
+      If the value is `AllNamespaces`, then the Operator will watch all namespaces (cluster-scoped Operators).
+      If the value is `SingleNamespace`, then the Operator will watch a namespace, not necessarily its own.
+      This is an optional parameter, but if the CSV does not support AllNamespaces, then this parameter becomes 
+      required to instruct run bundle with the appropriate target namespace.
+  - name: OO_INDEX_IMG
+    default: "quay.io/operator-framework/upstream-opm-builder:latest"
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply ci-operator/step-registry/optional-operators/ci/run-bundle/optional-operators-ci-run-bundle-ref.yaml:30]: In conversation it looks like we need to dynamically resolve this by connecting to the cluster under test, determining its' version, and choosing the appropriate image
